### PR TITLE
Improve performance of commands for Foxy on Windows by 4x

### PR DESCRIPTION
**Pull request by Ace314159**
This improves the performance of commands like `ros2 topic list`, `ros2 service list`, `ros2 node list`, and I assume anything else using the daemon proxy. I was able to find this change and an explanation from [this Stack Overflow post](https://stackoverflow.com/questions/14504450/pythons-xmlrpc-extremely-slow-one-second-per-call).

When having no nodes running, the execution time of the above commands go from 5.3 seconds to 1.2 seconds on my machine. This is still much slower than running these commands in WSL, but now the majority of the delay seems to be from the imports in `ros2-script.py` and `list.py`

This also addresses #637

### CODE REVIEW

**Review by audrow — COMMENTED** [review]
This seems like a good change to me. Here are a few thoughts:

- It seems that a few tests are failing in the PR job - maybe we can use `127.0.0.1` only for Windows platforms? Perhaps that would make the tests pass.
- Should this be retargeted to rolling and then maybe backported to Foxy (and Galactic) as a bug fix?
- Is there a good way to test this? I believe some tests are disabled on Windows for speed. Maybe some of those can be enabled?

**Comment by clalancette** [comment]
> * Should this be retargeted to rolling and then maybe backported to Foxy (and Galactic) as a bug fix?

Yes, we should definitely retarget it to rolling, as that is our standard development process.  If we find we can make it work on Rolling, then we can consider backporting.

**Comment by Ace314159** [comment]
I'm not able to reproduce these the test failures on WSL, and this change already seems to be made in rolling [here](https://github.com/ros2/ros2cli/blob/master/ros2cli/ros2cli/daemon/__init__.py#L40), which seems to be fine. I can make the same change to galactic once the test failures on Foxy are figured out.

Are these test known to be flaky or is it actually the change itself causing the failures?

I can try enabling the tests on Windows to test this. The commands are still about 4x slower than in WSL, so it still might not be fast enough to pass all the tests.

**Comment by Ace314159** [comment]
In [Build 67](https://build.ros2.org/job/Fpr__ros2cli__ubuntu_focal_amd64/67/#showFailuresLink), service and doctor which originally failed don't fail, but topic and node still do. In [Build 68](https://build.ros2.org/job/Fpr__ros2cli__ubuntu_focal_amd64/68/#showFailuresLink), doctor doesn't fail, but action does now, and topic and node still fail.

It appears that only topic and node consistently fail, but I still can't reproduce either of them locally on WSL. Any ideas why it's failing here even though the same change is already in rolling?

Also, can you point me to where I can see the Windows PR job?

**Comment by Ace314159** [comment]
Even when using `127.0.0.1` only on Windows, the tests still don't pass. Any ideas for what else I could try?

**Comment by clalancette** [comment]
Especially for Foxy, let's keep the `test_cli` tests disabled for Windows.  We had massive problems with them, and I'm pretty sure fixing them isn't going to be as simple as changing to 127.0.0.1 here.

**Comment by Ace314159** [comment]
@clalancette Is there anything else you'd like me to do? Although this change alone might not be enough to fix the tests, this still results in a substantial speed increase, making the ros2 CLI a bit more usable on Windows.

**Comment by Ace314159** [comment]
@audrow @clalancette Friendly ping

**Comment by Ace314159** [comment]
I was able to generate flamegraphs and inverted flamegraphs using [flameprof](https://github.com/baverman/flameprof/) in order to better visualize the impacts of this seemingly minimal change. I was also able to find identify another bottleneck involvinng `os.stat` that might be able to explain any remaining speed difference between Windows and Linux.

I ran `ros2 topic list` without any nodes running, before and after the changes on Windows. I also ran it on WSL to get an estimate of the performance on Linux to identify any remaining bottlenecks.

Before the change, the call took ~5.3 seconds, and the flame graph and inverted flame graph look like below:
image
image

The majority of the time is spent in the `create_connection` method of socket, which I assume is called when using the `ServerProxy`.

After the change, the call took ~1.2 seconds, and the flame graph and inverted flame graph look like below:
image
image

The massive bottleneck involving the `ServerProxy` is removed, and the remaining bottleneck is due to `nt.stat`. Reading from the inverted flame graph, the total time of all calls of `nt.stat` calls is ~0.29 seconds.

The WSL call took ~0.9 seconds, and the flame graph and inverted flame graph look like below:
image
image

In WSL, `posix.stat`, which is the largest (but still pretty small) dark purple bar on the right in the inverted flame graph, is much smaller and barely visible. Reading from the inverted flame graph, the total time of all calls of `posix.stat` is ~0.059 seconds.

This probably means that the ~0.3 second difference between WSL and Windows is due to `os.stat` being much slower on Windows.

**Comment by Ace314159** [comment]
@clalancette @audrow Just to note, since I seemed to have forgotten to mention this, this change is already in rolling, so it would just be a matter of backporting to foxy and galactic:
https://github.com/ros2/ros2cli/blob/0479d5d17ee0f7ccad2b30835d464d5aaa19bb65/ros2cli/ros2cli/daemon/__init__.py#L40

File: ros2cli/ros2cli/node/daemon.py [hunk 1]
```diff
@@ -100,8 +100,9 @@ class DaemonNode:
 
     def __init__(self, args):
         self._args = args
+        host = '127.0.0.1' if platform.system() == 'Windows' else 'localhost'
```
Comment (clalancette) [inline ros2cli/ros2cli/node/daemon.py:103]: So if we are going to follow in the same idea as what Rolling has, I'll suggest we just unconditionally use 127.0.0.1 here.

**Review by clalancette — CHANGES_REQUESTED** [review]
One small change to make, then I think this is good to go.  We'll also have to figure out why CI is unhappy.

**Review by clalancette — APPROVED** [review]
This looks like a reasonable enough backport for me.  I'll run CI on it next.

**Comment by clalancette** [comment]
OK, CI is happy (the failures on Windows are well-known, and not related to this PR).  I'm going to go ahead and merge this one in, thanks for all of the work here.

**Comment by jacobperron** [comment]
I noticed this was never ported to Galactic. I've opened up a PR here: https://github.com/ros2/ros2cli/pull/740

**Pull request merged**